lms/views/content_item_selection.py

```diff
@@ -1,90 +1,98 @@
-        "oauth_nonce": lti_params["oauth_nonce"],
-        "oauth_consumer_key": lti_params["oauth_consumer_key"],
-        "oauth_signature_method": lti_params["oauth_signature_method"],
-        "oauth_signature": lti_params["oauth_signature"],
-        "jwt_token": jwt,
-    }
-    # These fields appear in blackboard launches, but not in canvas
-    # launches
-    if "resource_link_id" in lti_params:
-        form_fields["resource_link_id"] = lti_params["resource_link_id"]
-    if "tool_consumer_instance_guid" in lti_params:
-        form_fields["tool_consumer_instance_guid"] = lti_params[
-            "tool_consumer_instance_guid"
-        ]
-
-    custom_lms_url = None
-    if "custom_canvas_api_domain" in lti_params:
-        custom_lms_url = lti_params["custom_canvas_api_domain"]
-
-    params = {
-        "content_item_return_url": content_item_return_url,
+def content_item_selection(request):
+    template_variables = {
+        # The URL that we'll POST the ContentItemSelection form submission
+        # (containing the user's selected document) to.
+        "content_item_return_url": request.params["content_item_return_url"],
+        # The "content item selection" that we submit to the
+        # content_item_return_url is actually an LTI launch URL with the
+        # selected document URL or file_id as a query parameter. To construct
+        # these launch URLs our JavaScript code needs the base URL of our LTI
+        # launch endpoint.
         "lti_launch_url": request.route_url("lti_launches"),
-        "form_fields": form_fields,
+        # The fields of the form that we'll POST to the content_item_return_url.
+        # (The JavaScript also adds the content item selection itself to the
+        # form as another field, in addition to the ones here.)
+        "form_fields": {
+            "lti_message_type": "ContentItemSelection",
+            "lti_version": request.params["lti_version"],
+        },
+        # Variables needed for initializing Google Picker.
         "google_client_id": request.registry.settings["google_client_id"],
         "google_developer_key": request.registry.settings["google_developer_key"],
-        "google_app_id": request.registry.settings["google_app_id"],
-        "lms_url": lms_url if custom_lms_url is None else custom_lms_url,
-        "api_url": request.route_url("canvas_proxy"),
-        "jwt": jwt,
     }
 
-    if canvas_files_available(request, params=lti_params):
-        params["course_id"] = lti_params["custom_canvas_course_id"]
-    return params
+    # Pass the URL of the LMS that is launching us to our JavaScript code.
+    # When we're being launched in an iframe within the LMS our JavaScript
+    # needs to pass this URL (which is the URL of the top-most page) to Google
+    # Picker, otherwise Picker refuses to launch inside an iframe.
+    if "custom_canvas_api_domain" in request.params:
+        template_variables["lms_url"] = request.params["custom_canvas_api_domain"]
```

Comment:
We should rename this to something more specific soon, in case someone else ends up trying to use it for a different purpose, eg. `top_frame_origin` or something.

I also wonder whether we might be able to get this value from somewhere else, such as the HTTP_REFERER header, rather than rely on the user to specify it correctly.

Reply:
Yeah. This whole variable is dodgy to be honest. Its value is either the unvalidated LMS domain entered by a user into https://lms.hypothes.is/welcome, or its the value of the `custom_canvas_api_domain` launch param which is the Canvas API's domain which is not quite the same thing as the domain of the top-level frame. In practice it seems to mostly work. If there's a better way to get the right value that would be good